java/src/main/java/com/google/appengine/tools/mapreduce/impl/shardedjob/ShardedJobRunner.java

```diff
@@ -386,16 +421,63 @@ public void runTask(final ShardedJobRunId jobId, final IncrementalTaskId taskId,
         }
 
         //taskState represents attempt of executing a slice of a shard of a sharded job
-        IncrementalTaskState taskState =
-          getAndValidateTaskState(lockAcquisition, taskId, sequenceNumber, jobState);
-        if (taskState == null) {
-          // some sort of error code happened
-
-          // seems like getAndValidationTaskState has potential side-effects, which need to be committed
-          lockAcquisition.commit();
-          return null;
+        IncrementalTaskState taskState = lookupTaskState(lockAcquisition, taskId);
+        IncrementalTaskStateValidationError validationError = validateTaskState(taskState, sequenceNumber, jobState);
+        if (validationError != null) {
+          switch (validationError) {
+            case TASK_GONE:
+              log.info(taskId + ": Task is gone, ignoring runTask call.");
+              lockAcquisition.commit();
+              //throw, so we re-try a few moments later and see if taskState has been updated
+              throw new IllegalStateException("Task is gone");
+
+            case TASK_NO_LONGER_ACTIVE:
+              log.info(taskId + ": Task is no longer active, ignoring runTask call.");
+              lockAcquisition.commit();
+              return null; //we're done here
+
+            case JOB_NO_LONGER_ACTIVE:
+              taskState.setStatus(new Status(StatusCode.ABORTED));
+              log.info(taskId + ": Job no longer active: " + jobState + ", aborting task.");
+
+              //TODO: has side-effects (enqueuing things)
+              updateTask(lockAcquisition, jobState, taskState, null, false);
```

Comment:
so this still has (potentially) has side-effects underneath; next step would be to split various invocations of 'updateTask' to more clearly explain the purpose (eg, `abortTask()` in this case??